File: core/src/test/java/net/typedrest/ElementEndpointTest.java
package net.typedrest;

import static com.github.tomakehurst.wiremock.client.WireMock.*;
import static com.github.tomakehurst.wiremock.stubbing.Scenario.STARTED;
import static org.apache.http.HttpHeaders.*;
import static org.apache.http.HttpStatus.*;
import static org.hamcrest.CoreMatchers.*;
import org.hamcrest.CoreMatchers;
import static org.hamcrest.MatcherAssert.*;
import org.junit.*;

public class ElementEndpointTest extends AbstractEndpointTest {

    private ElementEndpoint<MockEntity> endpoint;

    @Before
    @Override
    public void before() {
        super.before();
        endpoint = new ElementEndpointImpl<>(entryEndpoint, "endpoint", MockEntity.class);
    }

    @Test
    public void testRead() throws Exception {
        stubFor(get(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withBody("{\"id\":5,\"name\":\"test\"}")));

        assertThat(endpoint.read(),
                is(equalTo(new MockEntity(5, "test"))));
    }

    @Test
    public void testReadCache() throws Exception {
        stubFor(get(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withHeader(ETAG, "\"123abc\"")
                        .withBody("{\"id\":5,\"name\":\"test\"}")));
        MockEntity result1 = endpoint.read();
        assertThat(result1, is(equalTo(new MockEntity(5, "test"))));

        stubFor(get(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .withHeader(IF_NONE_MATCH, equalTo("\"123abc\""))
                .willReturn(aResponse()
                        .withStatus(SC_NOT_MODIFIED)));
        MockEntity result2 = endpoint.read();
        assertThat(result2, is(equalTo(new MockEntity(5, "test"))));

        assertThat("Cache responses, not deserialized objects",
                result2, is(not(sameInstance(result1))));
    }

    @Test
    public void testExistsTrue() throws Exception {
        stubFor(head(urlEqualTo("/endpoint"))
                .willReturn(aResponse()
                        .withStatus(SC_OK)));

        assertThat(endpoint.exists(), is(true));
    }

    @Test
    public void testExistsFalse() throws Exception {
        stubFor(head(urlEqualTo("/endpoint"))
                .willReturn(aResponse()
                        .withStatus(SC_NOT_FOUND)));

        assertThat(endpoint.exists(), is(false));
    }

    @Test
    @Ignore("Works in isolation but fails when executed as part of test suite")
    public void testSetNoResult() throws Exception {
        stubFor(put(urlEqualTo("/endpoint"))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"test\"}"))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));

        endpoint.set(new MockEntity(5, "test"));
    }

    @Test
    @Ignore("Works in isolation but fails when executed as part of test suite")
    public void testSetResult() throws Exception {
        stubFor(put(urlEqualTo("/endpoint"))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"test\"}"))
                .willReturn(aResponse()
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withBody("{\"id\":5,\"name\":\"testXXX\"}")));

        assertThat(endpoint.set(new MockEntity(5, "test")),
                is(equalTo(new MockEntity(5, "testXXX"))));
    }

    @Test
    public void testSetETag() throws Exception {
        stubFor(get(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withHeader(ETAG, "\"123abc\"")
                        .withBody("{\"id\":5,\"name\":\"test\"}")));
        MockEntity entity = endpoint.read();

        stubFor(put(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .withHeader(IF_MATCH, matching("\"123abc\""))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"test\"}"))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));
        endpoint.set(entity);
    }

    @Test
    @Ignore("Works in isolation but fails when executed as part of test suite")
    public void testSetWithNullValue() throws Exception {
        stubFor(put(urlEqualTo("/endpoint"))
                .withRequestBody(equalToJson("{\"id\":5}"))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));

        endpoint.set(new MockEntity(5, null));
    }

    @Test
    public void testUpdateRetry() throws Exception {
        stubFor(get(urlEqualTo("/endpoint")).inScenario("testUpdateRetry")
                .whenScenarioStateIs(STARTED)
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withHeader(ETAG, "\"1\"")
                        .withBody("{\"id\":5,\"name\":\"test1\"}"))
                .willSetStateTo("Second attempt"));
        stubFor(put(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .withHeader(IF_MATCH, matching("\"1\""))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"testX\"}"))
                .willReturn(aResponse()
                        .withStatus(SC_PRECONDITION_FAILED)));
        stubFor(get(urlEqualTo("/endpoint")).inScenario("testUpdateRetry")
                .whenScenarioStateIs("Second attempt")
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withHeader(ETAG, "\"2\"")
                        .withBody("{\"id\":5,\"name\":\"test2\"}")));
        stubFor(put(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .withHeader(IF_MATCH, matching("\"2\""))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"testX\"}"))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));

        endpoint.update(x -> x.setName("testX"));
    }

    @Test(expected = IllegalStateException.class)
    public void testUpdateFail() throws Exception {
        stubFor(get(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withHeader(ETAG, "\"1\"")
                        .withBody("{\"id\":5,\"name\":\"test1\"}")));
        stubFor(put(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .withHeader(IF_MATCH, matching("\"1\""))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"testX\"}"))
                .willReturn(aResponse()
                        .withStatus(SC_PRECONDITION_FAILED)));

        endpoint.update(x -> x.setName("testX"), 0);
    }

    @Test
    @Ignore("Works in isolation but fails when executed as part of test suite")
    public void testMergeNoResult() throws Exception {
        stubFor(put(urlEqualTo("/endpoint"))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"test\"}"))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));

        endpoint.merge(new MockEntity(5, "test"));
    }

    @Test
    @Ignore("Works in isolation but fails when executed as part of test suite")
    public void testMergeResult() throws Exception {
        stubFor(put(urlEqualTo("/endpoint"))
                .withRequestBody(equalToJson("{\"id\":5,\"name\":\"test\"}"))
                .willReturn(aResponse()
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withBody("{\"id\":5,\"name\":\"testXXX\"}")));

        assertThat(endpoint.merge(new MockEntity(5, "test")),
                is(equalTo(new MockEntity(5, "testXXX"))));
    }

    @Test
    public void testDelete() throws Exception {
        stubFor(delete(urlEqualTo("/endpoint"))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));

        endpoint.delete();
    }

    @Test
    public void testDeleteETag() throws Exception {
        stubFor(get(urlEqualTo("/endpoint"))
                .withHeader(ACCEPT, equalTo(JSON_MIME))
                .willReturn(aResponse()
                        .withStatus(SC_OK)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withHeader(ETAG, "\"123abc\"")
                        .withBody("{\"id\":5,\"name\":\"test\"}")));
        endpoint.read();

        stubFor(delete(urlEqualTo("/endpoint"))
                .withHeader(IF_MATCH, matching("\"123abc\""))
                .willReturn(aResponse()
                        .withStatus(SC_NO_CONTENT)));

        endpoint.delete();
    }

    @Test
    public void testException() throws Exception {
        stubFor(get(urlEqualTo("/endpoint"))
                .willReturn(aResponse()
                        .withStatus(SC_BAD_REQUEST)
                        .withHeader(CONTENT_TYPE, JSON_MIME)
                        .withBody("{\"message\":\"test\"}")));

        String exceptionMessage = null;
        try {
            endpoint.read();
        } catch (IllegalArgumentException ex) {
            exceptionMessage = ex.getMessage();
        }

        assertThat(exceptionMessage, is(CoreMatchers.equalTo("test")));
    }
}

File: typedrest-serializers-jackson/src/test/kotlin/net/typedrest/endpoints/generic/ElementEndpointTest.kt
package net.typedrest.endpoints.generic

import net.typedrest.MockEntity
import net.typedrest.endpoints.AbstractEndpointTest
import net.typedrest.http.*
import net.typedrest.tests.*
import okhttp3.mockwebserver.MockResponse
import kotlin.test.*

class ElementEndpointTest : AbstractEndpointTest() {
    private val endpoint = ElementEndpointImpl(entryEndpoint, "endpoint", MockEntity::class.java)

    @Test
    fun testRead() {
        server.enqueue(
            MockResponse()
                .setJsonBody("""{"id":5,"name":"test"}""")
        )

        assertEquals(MockEntity(5, "test"), endpoint.read())
    }

    @Test
    fun testReadCustomMimeWithJsonSuffix() {
        server.enqueue(
            MockResponse()
                .setJsonBody("""{"id":5,"name":"test"}""")
        )

        assertEquals(MockEntity(5, "test"), endpoint.read())
    }

    @Test
    fun testReadCacheETag() {
        server.enqueue(
            MockResponse()
                .setHeader("ETag", "\"123abc\"")
                .setJsonBody("""{"id":5,"name":"test"}""")
        )
        val result1 = endpoint.read()
        server.assertRequest(HttpMethod.GET)
        assertEquals(MockEntity(5, "test"), result1)

        server.enqueue(MockResponse().setResponseCode(HttpStatusCode.NotModified))
        val result2 = endpoint.read()
        server.assertRequest(HttpMethod.GET).withHeader("If-None-Match", "\"123abc\"")
        assertEquals(MockEntity(5, "test"), result2)
        assertNotSame(result1, result2, message = "Cache responses, not deserialized objects")
    }

    @Test
    fun testSetResult() {
        server.enqueue(MockResponse().setJsonBody("""{"id":5,"name":"testXXX"}"""))

        assertEquals(MockEntity(5, "testXXX"), endpoint.set(MockEntity(5, "test")))
        server.assertRequest(HttpMethod.PUT)
            .withJsonBody("""{"id":5,"name":"test"}""")
    }

    @Test
    fun testSetNoResult() {
        server.enqueue(MockResponse().setResponseCode(HttpStatusCode.NoContent))

        assertNull(endpoint.set(MockEntity(5, "test")))
        server.assertRequest(HttpMethod.PUT)
            .withJsonBody("""{"id":5,"name":"test"}""")
    }

    @Test
    fun testSetETag() {
        server.enqueue(
            MockResponse()
                .setHeader("ETag", "\"123abc\"")
                .setJsonBody("""{"id":5,"name":"test"}""")
        )
        val result = endpoint.read()
        server.assertRequest(HttpMethod.GET)

        server.enqueue(MockResponse().setResponseCode(HttpStatusCode.NoContent))
        endpoint.set(result)
        server.assertRequest(HttpMethod.PUT)
            .withHeader("If-Match", "\"123abc\"")
            .withJsonBody("""{"id":5,"name":"test"}""")
    }

    @Test
    fun testUpdateRetry() {
        server.enqueue(MockResponse().setJsonBody("""{"id":5,"name":"test1"}""").addHeader("ETag", "\"1\""))
        server.enqueue(MockResponse().setResponseCode(HttpStatusCode.PreconditionFailed))
        server.enqueue(MockResponse().setJsonBody("""{"id":5,"name":"test2"}""").addHeader("ETag", "\"2\""))
        server.enqueue(MockResponse().setResponseCode(HttpStatusCode.NoContent))

        endpoint.update({ MockEntity(it.id, "testX") })

        server.assertRequest(HttpMethod.GET)
        server.assertRequest(HttpMethod.PUT).withJsonBody("""{"id":5,"name":"testX"}""").withHeader("If-Match", "\"1\"")
        server.assertRequest(HttpMethod.GET)
        server.assertRequest(HttpMethod.PUT).withJsonBody("""{"id":5,"name":"testX"}""").withHeader("If-Match", "\"2\"")
    }

    @Test
    fun testMergeResult() {
        server.enqueue(MockResponse().setJsonBody("""{"id":5,"name":"testXXX"}"""))

        assertEquals(MockEntity(5, "testXXX"), endpoint.merge(MockEntity(5, "test")))
        server.assertRequest(HttpMethod.PATCH)
            .withJsonBody("""{"id":5,"name":"test"}""")
    }

    @Test
    fun testMergeNoResult() {
        server.enqueue(MockResponse().setResponseCode(HttpStatusCode.NoContent))

        assertNull(endpoint.merge(MockEntity(5, "test")))
        server.assertRequest(HttpMethod.PATCH)
            .withJsonBody("""{"id":5,"name":"test"}""")
    }
}
